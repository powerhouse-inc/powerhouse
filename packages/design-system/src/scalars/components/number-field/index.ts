export * from "./number-field.js";
