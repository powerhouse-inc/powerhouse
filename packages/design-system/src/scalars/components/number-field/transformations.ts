export type Transformation = (value: string | number) => string | number;
const validNumber = /^-?\d+$/;
export const applyTransformations = (
  value: string | number,
  transformations: Transformation[],
): string | number => {
  return transformations.reduce((acc, transform) => transform(acc), value);
};

export const applyBigInt = (value: string | number): string | number => {
  const stringValue = value.toString().replace(/[^-\d.]/g, "");
  const integerPart = stringValue.split(".")[0];
  return validNumber.test(integerPart) ? BigInt(integerPart).toString() : "";
};

export const applyMaxSafeInteger = (
  value: string | number,
): string | number => {
  const numericValue = parseFloat(value.toString());
  return numericValue > Number.MAX_SAFE_INTEGER
    ? value.toString().replace(/[^-\d.]/g, "")
    : value;
};

export const applyPrecision = (precision: number, trailingZeros = false) => {
  return (value: string | number): string | number => {
    const numericValue = parseFloat(value.toString());
    const formattedValue = numericValue.toFixed(precision);
    return trailingZeros
      ? formattedValue
      : parseFloat(formattedValue).toString();
  };
};
