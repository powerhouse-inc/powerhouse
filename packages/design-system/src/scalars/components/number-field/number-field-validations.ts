import { type NumberFieldProps } from "./number-field.js";

export const isInteger = (value: unknown): boolean =>
  Number.isInteger(Number(value));

export const isFloat = (value: unknown): boolean => {
  const number = Number(value);
  return !isNaN(number) && isFinite(number);
};

export const MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER;

// Utility function to check if a value is a valid number
export const isValidNumber = (value: unknown): boolean => {
  return /^-?\d+(\.\d+)?$/.test(value as string);
};

export const validateNumericType =
  ({ numericType }: NumberFieldProps) =>
  (value: unknown) => {
    if (value === "") {
      return true;
    }

    if (!isValidNumber(value)) {
      return "Value is not a valid number";
    }

    const isPositive = Number(value) > 0;
    const isNegative = Number(value) < 0;
    const isInt = isInteger(value);
    const numberValue = Number(value);

    if (
      Math.abs(numberValue) > Number.MAX_SAFE_INTEGER &&
      numericType !== "BigInt"
    ) {
      const isFloatType =
        numericType &&
        [
          "Float",
          "PositiveFloat",
          "NegativeFloat",
          "NonNegativeFloat",
          "NonPositiveFloat",
        ].includes(numericType);

      return isFloatType
        ? "Value is too large for float"
        : "Value is too large for int";
    }

    switch (numericType) {
      case "PositiveInt": {
        if (!isInt) return "Value must be a positive integer";
        if (isPositive) return true;
        return "Value must be a positive integer";
      }

      case "NegativeInt": {
        if (!isInt) return "Value must be a negative integer";
        if (Number(value) >= 0) return "Value must be a negative integer";
        if (isNegative) return true;
        return "Value must be a negative integer";
      }

      case "NonNegativeInt": {
        if (!isInt) return "Value must be a non-negative integer";
        if (Number(value) >= 0) return true;
        return "Value must be a non-negative integer";
      }

      case "NonPositiveInt": {
        if (!isInt) return "Value must be a non-positive integer";
        if (Number(value) <= 0) return true;
        return "Value must be a non-positive integer";
      }

      case "PositiveFloat": {
        if (!isFloat(value)) return "Value must be a positive float";

        if (parseFloat(value as string) > 0) return true;
        return "Value must be a positive float";
      }

      case "NegativeFloat": {
        if (!isFloat(value)) return "Value must be a negative float";
        if (parseFloat(value as string) >= 0)
          return "Value must be a negative float";
        if (isNegative) return true;
        return "Value must be a negative float";
      }

      case "NonNegativeFloat": {
        if (!isFloat(value)) return "Value must be a non-negative float ";

        if (parseFloat(value as string) >= 0) return true;
        return "Value must be a non-negative float";
      }

      case "NonPositiveFloat": {
        if (!isFloat(value)) return "Value must be a non-positive float";
        if (parseFloat(value as string) <= 0) return true;
        return "Value must be a non-positive float";
      }

      case "Int": {
        if (!isInt) return "Value must be an integer";
        return true;
      }

      case "Float": {
        if (!isFloat(value)) return "Value must be a float";
        return true;
      }

      case "BigInt": {
        if (isInt) return true;
        return "This is not a valid BigInt";
      }

      default: {
        return true;
      }
    }
  };
