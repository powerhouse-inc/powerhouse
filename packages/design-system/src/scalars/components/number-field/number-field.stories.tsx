import type { Meta, StoryObj } from "@storybook/react";
import { withForm } from "../../lib/decorators.js";
import {
  getDefaultArgTypes,
  getValidationArgTypes,
  PrebuiltArgTypes,
  StorybookControlCategory,
} from "../../lib/storybook-arg-types.js";
import { NumberField } from "./number-field.js";

const meta = {
  title: "Document Engineering/Scalars/Number Field",
  component: NumberField,
  parameters: {
    layout: "centered",
  },
  decorators: [withForm],
  tags: ["autodocs"],
  argTypes: {
    ...getDefaultArgTypes({
      valueControlType: "number",
      valueType: "number",
    }),
    step: {
      control: "number",
      description: "Step value for the input field",
      table: {
        type: { summary: "number" },
        defaultValue: { summary: "1" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    autoFocus: {
      table: {
        disable: true,
      },
    },
    numericType: {
      control: {
        type: "select",
      },
      options: [
        "PositiveInt",
        "NegativeInt",
        "NonNegativeInt",
        "NonPositiveInt",
        "PositiveFloat",
        "NegativeFloat",
        "NonNegativeFloat",
        "NonPositiveFloat",
        "BigInt",
        "Int",
        "Float",
      ],
      description:
        "Specifies the numeric type of the input field. Possible values are:\n\n\n" +
        "- PositiveInt: Positive whole numbers (1, 2, 3)\n" +
        "- NegativeInt: Negative whole numbers (-1, -2, -3)\n" +
        "- NonNegativeInt: Zero and positive whole numbers (0, 2)\n" +
        "- NonPositiveInt: Zero and negative whole numbers ( 0, -2)\n" +
        "- PositiveFloat: Positive decimals (1.0, 2.5)\n" +
        "- NegativeFloat: Negative decimals (-1.0, -2.5)\n" +
        "- NonNegativeFloat: Zero and positive decimals (0.0, 1.0)\n" +
        "- NonPositiveFloat: Zero and negative decimals (0.0, -1.0)\n" +
        "- BigInt: Large integers (9999999999999999)\n" +
        "- Float: Any decimal number (1.0, 2.5, -1.0, -2.5)\n" +
        "- Int: Any integer number, including positive (1, 2), negative (-1, -2), and zero\n",
      table: {
        defaultValue: { summary: "Float" },
        type: {
          summary:
            '"PositiveInt" | "NegativeInt" | "NonNegativeInt" | "NonPositiveInt" | "PositiveFloat" | "NegativeFloat" | "NonNegativeFloat" | "NonPositiveFloat" | "BigInt" | "Int" | "Float"',
        },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    ...PrebuiltArgTypes.placeholder,
    ...getValidationArgTypes(),
    ...PrebuiltArgTypes.minValue,
    ...PrebuiltArgTypes.maxValue,
    ...PrebuiltArgTypes.precision,
    ...PrebuiltArgTypes.trailingZeros,
  },
  args: {
    name: "number-field",
    errors: [],
    warnings: [],
  },
} satisfies Meta<typeof NumberField>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    name: "Label",
    label: "Label",
    placeholder: "Enter a number",
    value: 1234,
  },
};
export const Active: Story = {
  args: {
    name: "Label",
    label: "Label",
    autoFocus: true,
    value: 45,
    placeholder: "Enter a number",
  },
  parameters: {
    pseudo: { focus: true },
  },
};
export const Disable: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 1234,
    disabled: true,
  },
};
export const Required: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 345,
    required: true,
    placeholder: "A number is required",
  },
};
export const WithWarning: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 23,
    warnings: ["Warning message"],
    placeholder: "Enter  number",
  },
};
export const WithError: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 23,
    errors: ["Error message"],
    placeholder: "Enter a number",
  },
};
export const WithMultipleErrors: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 23,
    errors: [
      "Error message number 1",
      "Error message number 2",
      "Error message number 3",
      "Error message number 4",
    ],
    placeholder: "Enter a number",
  },
};

export const WithValue: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 23,
    placeholder: "Enter a number",
  },
};

export const WithDescription: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    description: "This is the field description",
    placeholder: "Enter a number",
  },
};

// Float Stories
export const WithFloatNumber: Story = {
  args: {
    label: "Label",
    name: "Label",
    value: 0.0,
    precision: 2,
    trailingZeros: true,
    placeholder: "Enter a decimal number",
  },
};
// BigInt Stories
export const WithBigInt: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 99992342343299,
    numericType: "BigInt",
    placeholder: "Enter a large number",
  },
};

// Step Stories
export const WithStep: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 456,
    step: 10,
    minValue: 20,
    placeholder: "Enter a number",
  },
};
