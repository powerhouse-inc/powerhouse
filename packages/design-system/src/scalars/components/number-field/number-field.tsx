import { Icon } from "#powerhouse";
import { cn } from "#scalars";
import { forwardRef, useId } from "react";
import { Input } from "../../../ui/components/index.js";
import { FormDescription } from "../fragments/form-description/index.js";
import { FormGroup } from "../fragments/form-group/index.js";
import { FormLabel } from "../fragments/form-label/index.js";
import { FormMessageList } from "../fragments/form-message/index.js";
import { withFieldValidation } from "../fragments/with-field-validation/index.js";
import { validateNumericType } from "./number-field-validations.js";
import type { InputNumberProps } from "./types.js";
import { useNumberField } from "./use-number-field.js";
import { regex } from "./utils.js";

export interface NumberFieldProps extends InputNumberProps {
  name: string;
  value?: number | bigint;
  defaultValue?: number | bigint;
  className?: string;
  pattern?: RegExp;
  onFocus?: (e: React.FocusEvent<HTMLInputElement>) => void;
}
export const NumberFieldRaw = forwardRef<HTMLInputElement, NumberFieldProps>(
  (
    {
      label,
      name,
      description,
      value,
      defaultValue,
      onChange,
      onBlur,
      errors,
      warnings,
      className,
      id: propId,
      minValue,
      maxValue,
      step = 1,
      pattern,
      trailingZeros = false,
      numericType = "Float",
      precision = 0,
      onFocus,
      ...props
    },
    ref,
  ) => {
    const generatedId = useId();
    const id = propId ?? generatedId;
    const {
      canIncrement,
      canDecrement,
      showSteps,
      preventInvalidCharsAndHandleArrows,
      stepValueHandler,
      blockInvalidPaste,
      preventLetterInput,
      isBigInt,
      handleBlur,

      handleFocus,
      buttonRef,
    } = useNumberField({
      value,
      maxValue,
      minValue,
      step,
      onChange,
      numericType,
      onBlur,
      trailingZeros,
      precision,
      onFocus,
    });

    return (
      <FormGroup>
        {label && (
          <FormLabel
            htmlFor={id}
            required={props.required}
            disabled={props.disabled}
            hasError={!!errors?.length}
            className="mb-[3px]"
          >
            {label}
          </FormLabel>
        )}
        <div className="relative flex items-center">
          <Input
            id={id}
            onFocus={handleFocus}
            name={name}
            className={cn("pr-8", className)}
            pattern={isBigInt ? regex.toString() : pattern?.toString()}
            type="text"
            inputMode="numeric"
            role="spinbutton"
            min={minValue}
            max={maxValue}
            aria-valuemin={minValue}
            aria-valuemax={maxValue}
            aria-invalid={!!errors?.length}
            onKeyDown={(e) => {
              preventLetterInput(e);
              preventInvalidCharsAndHandleArrows(e);
            }}
            value={value === undefined ? "" : value.toString()}
            onBlur={handleBlur}
            defaultValue={defaultValue?.toString()}
            onChange={onChange}
            onPaste={blockInvalidPaste}
            ref={ref}
            data-cast={isBigInt ? "BigInt" : "Number"}
            {...props}
          />
          {showSteps && (
            <div className="absolute inset-y-2 right-3 flex flex-col justify-center">
              <button
                aria-label="Increment"
                disabled={canIncrement}
                onMouseDown={(e) => e.preventDefault()}
                type="button"
                onClick={(e) => {
                  stepValueHandler(e, "increment");
                  if (buttonRef.current) {
                    buttonRef.current.focus();
                  }
                }}
              >
                <Icon
                  size={10}
                  name="ChevronDown"
                  className={cn(
                    "rotate-180 text-gray-700 dark:text-gray-300",
                    canIncrement && "cursor-not-allowed",
                  )}
                />
              </button>
              <button
                aria-label="Decrement"
                onMouseDown={(e) => e.preventDefault()}
                disabled={canDecrement}
                type="button"
                onClick={(e) => {
                  stepValueHandler(e, "decrement");
                  if (buttonRef.current) {
                    buttonRef.current.focus();
                  }
                }}
              >
                <Icon
                  size={10}
                  name="ChevronDown"
                  className={cn(
                    "items-center justify-center text-gray-700 dark:text-gray-300",
                    canDecrement && "cursor-not-allowed",
                  )}
                />
              </button>
            </div>
          )}
        </div>
        {description && <FormDescription>{description}</FormDescription>}
        {warnings && <FormMessageList messages={warnings} type="warning" />}
        {errors && <FormMessageList messages={errors} type="error" />}
      </FormGroup>
    );
  },
);

export const NumberField = withFieldValidation<NumberFieldProps>(
  NumberFieldRaw,
  {
    validations: {
      _numericType: validateNumericType,
    },
  },
);
NumberField.displayName = "NumberField";
