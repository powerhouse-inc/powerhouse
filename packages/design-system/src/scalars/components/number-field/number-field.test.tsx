import { Button } from "#powerhouse";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { renderWithForm } from "../../lib/testing.js";
import { Form } from "../form/form.js";
import { NumberField } from "./number-field.js";

describe("NumberField", () => {
  const mockOnChange = vi.fn();

  beforeEach(() => {
    mockOnChange.mockClear(); // Limpia el mock antes de cada prueba
  });

  it("should match snapshot", () => {
    const { container } = renderWithForm(
      <NumberField
        label="Test Label"
        name="Label"
        value={345}
        trailingZeros
        precision={2}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it("should pass the correct BigInt value on form submission", async () => {
    const mockOnSubmit = vi.fn();
    const user = userEvent.setup();

    render(
      <Form onSubmit={mockOnSubmit}>
        {({ formState: { isSubmitting } }) => (
          <div>
            <NumberField
              label="BigInt Field"
              name="BigInt Field"
              numericType="BigInt"
            />
            <Button type="submit" disabled={isSubmitting}>
              {isSubmitting ? "Submitting..." : "Submit"}
            </Button>
          </div>
        )}
      </Form>,
    );

    // We need to mock the data beforehand to avoid loss of precision
    const bigIntString = "9007199254740993";
    const bigIntValue = BigInt(bigIntString);

    // Simulate user input by directly setting the value
    const input = screen.getByLabelText("BigInt Field");
    fireEvent.change(input, { target: { value: bigIntString } });
    fireEvent.blur(input);

    // Submit the form
    const submitButton = screen.getByText("Submit");
    await user.click(submitButton);
    await waitFor(() => {
      expect(mockOnSubmit).toHaveBeenCalledTimes(1);
      expect(mockOnSubmit).toHaveBeenCalledWith({
        "BigInt Field": bigIntValue,
      });
    });
  });

  it("should render error messages when provided", () => {
    renderWithForm(
      <NumberField
        label="Test Label"
        onChange={mockOnChange}
        errors={["Error 1", "Error 2"]}
        name="Label"
      />,
    );
    expect(screen.getByText("Test Label")).toBeInTheDocument();
    expect(screen.getByText("Test Label")).toBeInTheDocument();
  });

  it("should pass the correct value on form submission", async () => {
    const mockOnSubmit = vi.fn();
    const user = userEvent.setup();

    // Render a form with NumberField
    renderWithForm(
      <Form onSubmit={mockOnSubmit}>
        {({ formState: { isSubmitting } }) => (
          <div className="flex w-[400px] flex-col gap-4">
            <NumberField label="Number Field" name="Number Field" />
            <Button type="submit" disabled={isSubmitting}>
              {isSubmitting ? "Submitting..." : "Submit"}
            </Button>
          </div>
        )}
      </Form>,
    );

    // Simulate user input
    const input = screen.getByLabelText("Number Field");
    await user.type(input, "42");

    // Submit the form
    const submitButton = screen.getByText("Submit");
    await user.click(submitButton);

    // Assert that the value is passed correctly
    expect(mockOnSubmit).toHaveBeenCalledWith({
      "Number Field": 42,
    });
  });

  it("should render label when label is provided", () => {
    renderWithForm(
      <NumberField label="Test Label" onChange={mockOnChange} name="Label" />,
    );
    expect(screen.getByLabelText("Test Label")).toBeInTheDocument();
  });

  it("should render the description when provided", () => {
    renderWithForm(
      <NumberField
        label="Test Label"
        description="This is a description"
        onChange={mockOnChange}
        name="Label"
      />,
    );
    expect(screen.getByText("This is a description")).toBeInTheDocument();
  });

  it("should render error messages when provided", () => {
    renderWithForm(
      <NumberField
        label="Test Label"
        onChange={mockOnChange}
        errors={["Error 1", "Error 2"]}
        name="Label"
      />,
    );
    expect(screen.getByText("Test Label")).toBeInTheDocument();
    expect(screen.getByText("Test Label")).toBeInTheDocument();
  });

  it("should render warning messages when provided", () => {
    renderWithForm(
      <NumberField
        label="Test Label"
        onChange={mockOnChange}
        warnings={["Warning 1", "Warning 2"]}
        name="Label"
      />,
    );
    expect(screen.getByText("Warning 1")).toBeInTheDocument();
    expect(screen.getByText("Warning 2")).toBeInTheDocument();
  });

  it("should call onChange when the input value changes", async () => {
    const user = userEvent.setup();
    renderWithForm(
      <NumberField label="Test Label" name="Label" onChange={mockOnChange} />,
    );
    const input = screen.getByLabelText("Test Label");

    await user.type(input, "10");
    expect(mockOnChange).toHaveBeenCalled();
    expect(input).toHaveValue("10");
  });

  it("should disable the input when disabled prop is true", () => {
    renderWithForm(
      <NumberField
        label="Test Label"
        onChange={mockOnChange}
        disabled
        name="Label"
      />,
    );
    const input = screen.getByLabelText("Test Label");

    expect(input).toBeDisabled();
  });

  it("should show the input as required when required prop is true", () => {
    renderWithForm(
      <NumberField
        label="Test Label"
        onChange={mockOnChange}
        required
        name="Label"
      />,
    );
    expect(screen.getByRole("spinbutton")).toHaveAttribute("required");
  });

  it("should support a defaultValue prop", () => {
    renderWithForm(
      <NumberField
        label="Test Label"
        defaultValue={5}
        onChange={mockOnChange}
        name="Label"
      />,
    );
    const input = screen.getByRole("spinbutton");

    expect(input).toHaveValue("5");
  });
  // Test for step
  it("should increment the value when increment arrow button is clicked", async () => {
    const user = userEvent.setup();

    renderWithForm(
      <NumberField
        label="Test Label"
        name="Label"
        value={5}
        step={1}
        onChange={mockOnChange}
      />,
    );

    const input = screen.getByRole("spinbutton");
    await user.click(input); // Simula el clic en el input

    // Ensure that the input has focus
    expect(input).toHaveFocus();

    const incrementButton = screen.getByRole("button", { name: /Increment/i });
    await user.click(incrementButton);

    // Verify that `mockOnChange` was called and that the input remains focused
    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(input).toHaveFocus();
    const eventArg = mockOnChange.mock
      .calls[0][0] as React.ChangeEvent<HTMLInputElement>;

    expect(eventArg.target).toMatchObject({
      value: 6,
    });
  });

  it("should decrement value when decrement button is clicked", async () => {
    const user = userEvent.setup();
    renderWithForm(
      <NumberField
        label="Test Label"
        name="Label"
        value={10}
        step={2}
        onChange={mockOnChange}
      />,
    );
    const input = screen.getByRole("spinbutton");
    await user.click(input);
    // Ensure that the input has focus
    expect(input).toHaveFocus();

    const decrementButton = screen.getByRole("button", { name: /Decrement/i });
    await user.click(decrementButton);
    // Ensure that the input has focus
    expect(mockOnChange).toHaveBeenCalledTimes(1);

    const eventArg = mockOnChange.mock
      .calls[0][0] as React.ChangeEvent<HTMLInputElement>;

    expect(eventArg.target).toMatchObject({
      value: 8,
    });
  });

  it("should not exceed maxValue when increment button is clicked", async () => {
    const user = userEvent.setup();
    renderWithForm(
      <NumberField
        label="Test Label"
        name="Label"
        value={10}
        maxValue={10}
        step={1}
        onChange={mockOnChange}
      />,
    );
    const input = screen.getByRole("spinbutton");
    await user.click(input);
    // Ensure that the input has focus
    expect(input).toHaveFocus();

    const incrementButton = screen.getByRole("button", { name: /Increment/i });
    await user.click(incrementButton);
    // // Aseg√∫rate de que el input tiene foco
    expect(mockOnChange).toHaveBeenCalledTimes(1);

    const eventArg = mockOnChange.mock
      .calls[0][0] as React.ChangeEvent<HTMLInputElement>;

    expect(eventArg.target.value).toBe("10");
  });

  it("should not go below minValue when decrement button is clicked", async () => {
    const user = userEvent.setup();
    renderWithForm(
      <NumberField
        label="Test Label"
        name="Label"
        value={1}
        minValue={1}
        step={1}
        onChange={mockOnChange}
      />,
    );
    const input = screen.getByRole("spinbutton");
    await user.click(input);
    // Ensure that the input has focus
    expect(input).toHaveFocus();

    const decrementButton = screen.getByRole("button", { name: /Decrement/i });
    await user.click(decrementButton);
    expect(mockOnChange).toHaveBeenCalledTimes(1);
    const eventArg = mockOnChange.mock
      .calls[0][0] as React.ChangeEvent<HTMLInputElement>;

    expect(eventArg.target.value).toBe("1");
  });

  //New test for the issues
  it("should handle NonPositiveFloat numeric type correctly", async () => {
    const mockOnSubmit = vi.fn();
    const user = userEvent.setup();

    render(
      <Form onSubmit={mockOnSubmit}>
        {({ formState: { isSubmitting } }) => (
          <div className="flex w-[400px] flex-col gap-4">
            <NumberField
              label="Float Field"
              name="floatField"
              numericType="NonPositiveFloat"
              precision={2}
            />
            <Button type="submit" disabled={isSubmitting}>
              Submit
            </Button>
          </div>
        )}
      </Form>,
    );

    const input = screen.getByLabelText("Float Field");
    await user.type(input, "-0.90");

    const submitButton = screen.getByText("Submit");
    await user.click(submitButton);

    expect(mockOnSubmit).toHaveBeenCalledWith({
      floatField: -0.9,
    });
  });

  it("should handle NegativeInt numeric type correctly", async () => {
    const mockOnSubmit = vi.fn();
    const user = userEvent.setup();

    render(
      <Form onSubmit={mockOnSubmit}>
        {({ formState: { isSubmitting } }) => (
          <div className="flex w-[400px] flex-col gap-4">
            <NumberField
              label="Negative Int Field"
              name="negativeField"
              numericType="NegativeInt"
            />
            <Button type="submit" disabled={isSubmitting}>
              Submit
            </Button>
          </div>
        )}
      </Form>,
    );

    const input = screen.getByLabelText("Negative Int Field");
    await user.type(input, "-87");

    const submitButton = screen.getByText("Submit");
    await user.click(submitButton);

    expect(mockOnSubmit).toHaveBeenCalledWith({
      negativeField: -87,
    });
  });

  it("should show placeholder when value is empty", async () => {
    const placeholder = "Enter a number";
    const user = userEvent.setup();

    renderWithForm(
      <NumberField
        label="Number Field"
        name="numberField"
        placeholder={placeholder}
      />,
    );

    const input = screen.getByLabelText("Number Field");
    await user.clear(input);

    expect(input).toHaveAttribute("placeholder", placeholder);
    expect(input).toHaveValue("");
  });

  it("should increment/decrement by specified step using keyboard", async () => {
    const user = userEvent.setup();
    const mockOnChange = vi.fn();

    renderWithForm(
      <NumberField
        label="Step Field"
        name="stepField"
        value={10}
        step={2}
        onChange={mockOnChange}
      />,
    );

    const input = screen.getByLabelText("Step Field");
    await user.type(input, "{arrowup}");

    expect(mockOnChange).toHaveBeenLastCalledWith(
      expect.objectContaining({
        target: { value: 12 },
      }),
    );

    await user.type(input, "{arrowdown}");

    expect(mockOnChange).toHaveBeenLastCalledWith(
      expect.objectContaining({
        target: { value: 10 },
      }),
    );
  });

  it("should handle precision correctly for decimal numbers", async () => {
    const mockOnSubmit = vi.fn();
    const user = userEvent.setup();

    render(
      <Form onSubmit={mockOnSubmit}>
        {({ formState: { isSubmitting } }) => (
          <div className="flex w-[400px] flex-col gap-4">
            <NumberField
              label="Precision Field"
              name="precisionField"
              precision={2}
            />
            <Button type="submit" disabled={isSubmitting}>
              Submit
            </Button>
          </div>
        )}
      </Form>,
    );

    const input = screen.getByLabelText("Precision Field");
    await user.type(input, "0.9");

    const submitButton = screen.getByText("Submit");
    await user.click(submitButton);

    expect(mockOnSubmit).toHaveBeenCalledWith({
      precisionField: 0.9,
    });
  });

  it("should accept integer values when numericType is PositiveFloat", async () => {
    const mockOnSubmit = vi.fn();
    const user = userEvent.setup();

    render(
      <Form onSubmit={mockOnSubmit}>
        {({ formState: { isSubmitting } }) => (
          <div>
            <NumberField
              label="Positive Float Field"
              name="positiveFloatField"
              numericType="PositiveFloat"
            />
            <Button type="submit" disabled={isSubmitting}>
              Submit
            </Button>
          </div>
        )}
      </Form>,
    );

    const input = screen.getByLabelText("Positive Float Field");
    await user.type(input, "42");

    const submitButton = screen.getByText("Submit");
    await user.click(submitButton);

    expect(mockOnSubmit).toHaveBeenCalledWith({
      positiveFloatField: 42,
    });
  });

  it("should retain the input value '4546-56' and display an error message", async () => {
    const mockOnSubmit = vi.fn();
    // const user = userEvent.setup({ delay: 100 });
    const user = userEvent.setup();

    render(
      <Form onSubmit={mockOnSubmit}>
        {({ formState: { isSubmitting } }) => (
          <div>
            <NumberField
              label="Special Case Field"
              name="specialCaseField"
              onChange={mockOnChange}
            />
            <Button type="submit" disabled={isSubmitting}>
              Submit
            </Button>
          </div>
        )}
      </Form>,
    );

    const input = screen.getByLabelText("Special Case Field");
    await user.type(input, "4546-56");

    expect(input).toHaveValue("4546-56");

    const submitButton = screen.getByText("Submit");
    await user.click(submitButton);

    expect(mockOnSubmit).not.toHaveBeenCalled();
  });

  //New test for the issues
  it("should handle NonPositiveFloat numeric type correctly", async () => {
    const mockOnSubmit = vi.fn();
    const user = userEvent.setup();

    render(
      <Form onSubmit={mockOnSubmit}>
        {({ formState: { isSubmitting } }) => (
          <div className="flex w-[400px] flex-col gap-4">
            <NumberField
              label="Float Field"
              name="floatField"
              numericType="NonPositiveFloat"
              precision={2}
            />
            <Button type="submit" disabled={isSubmitting}>
              Submit
            </Button>
          </div>
        )}
      </Form>,
    );

    const input = screen.getByLabelText("Float Field");
    await user.type(input, "-0.90");

    const submitButton = screen.getByText("Submit");
    await user.click(submitButton);

    expect(mockOnSubmit).toHaveBeenCalledWith({
      floatField: -0.9,
    });
  });

  it("should handle NegativeInt numeric type correctly", async () => {
    const mockOnSubmit = vi.fn();
    const user = userEvent.setup();

    render(
      <Form onSubmit={mockOnSubmit}>
        {({ formState: { isSubmitting } }) => (
          <div className="flex w-[400px] flex-col gap-4">
            <NumberField
              label="Negative Int Field"
              name="negativeField"
              numericType="NegativeInt"
            />
            <Button type="submit" disabled={isSubmitting}>
              Submit
            </Button>
          </div>
        )}
      </Form>,
    );

    const input = screen.getByLabelText("Negative Int Field");
    await user.type(input, "-87");

    const submitButton = screen.getByText("Submit");
    await user.click(submitButton);

    expect(mockOnSubmit).toHaveBeenCalledWith({
      negativeField: -87,
    });
  });

  it("should show placeholder when value is empty", async () => {
    const placeholder = "Enter a number";
    const user = userEvent.setup();

    renderWithForm(
      <NumberField
        label="Number Field"
        name="numberField"
        placeholder={placeholder}
      />,
    );

    const input = screen.getByLabelText("Number Field");
    await user.clear(input);

    expect(input).toHaveAttribute("placeholder", placeholder);
  });

  it("should increment/decrement by specified step using keyboard", async () => {
    const user = userEvent.setup();
    const mockOnChange = vi.fn();

    renderWithForm(
      <NumberField
        label="Step Field"
        name="stepField"
        value={10}
        step={2}
        onChange={mockOnChange}
      />,
    );

    const input = screen.getByLabelText("Step Field");
    await user.type(input, "{arrowup}");

    expect(mockOnChange).toHaveBeenLastCalledWith(
      expect.objectContaining({
        target: { value: 12 },
      }),
    );

    await user.type(input, "{arrowdown}");

    expect(mockOnChange).toHaveBeenLastCalledWith(
      expect.objectContaining({
        target: { value: 10 },
      }),
    );
  });

  it("should handle precision correctly for decimal numbers", async () => {
    const mockOnSubmit = vi.fn();
    const user = userEvent.setup();

    render(
      <Form onSubmit={mockOnSubmit}>
        {({ formState: { isSubmitting } }) => (
          <div className="flex w-[400px] flex-col gap-4">
            <NumberField
              label="Precision Field"
              name="precisionField"
              precision={2}
            />
            <Button type="submit" disabled={isSubmitting}>
              Submit
            </Button>
          </div>
        )}
      </Form>,
    );

    const input = screen.getByLabelText("Precision Field");
    await user.type(input, "0.9");

    const submitButton = screen.getByText("Submit");
    await user.click(submitButton);

    expect(mockOnSubmit).toHaveBeenCalledWith({
      precisionField: 0.9,
    });
  });

  it("should accept integer values when numericType is PositiveFloat", async () => {
    const mockOnSubmit = vi.fn();
    const user = userEvent.setup();

    render(
      <Form onSubmit={mockOnSubmit}>
        {({ formState: { isSubmitting } }) => (
          <div>
            <NumberField
              label="Positive Float Field"
              name="positiveFloatField"
              numericType="PositiveFloat"
            />
            <Button type="submit" disabled={isSubmitting}>
              Submit
            </Button>
          </div>
        )}
      </Form>,
    );

    const input = screen.getByLabelText("Positive Float Field");
    await user.type(input, "42");

    const submitButton = screen.getByText("Submit");
    await user.click(submitButton);

    expect(mockOnSubmit).toHaveBeenCalledWith({
      positiveFloatField: 42,
    });
  });
});
