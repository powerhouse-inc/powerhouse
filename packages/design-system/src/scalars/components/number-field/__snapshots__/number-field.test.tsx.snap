// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NumberField > should match snapshot 1`] = `
<div>
  <form
    id=":r0:"
    novalidate=""
  >
    <div
      class="flex flex-col gap-2"
    >
      <label
        class="inline-flex items-center text-sm font-medium leading-4 text-gray-900 dark:text-gray-50 mb-[3px]"
        for=":r1:"
      >
        Test Label
      </label>
      <div
        class="relative flex items-center"
      >
        <input
          aria-invalid="false"
          class="flex h-9 w-full rounded-md text-sm font-normal leading-5 text-gray-900 dark:text-gray-50 dark:border-charcoal-700 dark:bg-charcoal-900 border border-gray-300 bg-white px-3 py-2 font-sans placeholder:text-gray-500 dark:placeholder:text-gray-600 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-0 focus-visible:ring-offset-white dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900 dark:focus:bg-charcoal-900 focus:bg-gray-50 disabled:cursor-not-allowed disabled:border-gray-300 disabled:bg-gray-50 disabled:text-gray-700 disabled:dark:border-charcoal-800 disabled:dark:bg-charcoal-900 disabled:dark:text-gray-300 pr-8"
          data-cast="Number"
          id=":r1:"
          inputmode="numeric"
          name="Label"
          role="spinbutton"
          type="text"
          value="345"
        />
      </div>
    </div>
  </form>
</div>
`;
